From dff4bf47f0015a5d76c5225ce8fa9b0b3433f276 Mon Sep 17 00:00:00 2001
Message-Id: <dff4bf47f0015a5d76c5225ce8fa9b0b3433f276.1427265330.git.chang-joon.lee@intel.com>
In-Reply-To: <b4c66aa6a04d2be0fca9fa4448c7188a581a57a5.1427265330.git.chang-joon.lee@intel.com>
References: <b4c66aa6a04d2be0fca9fa4448c7188a581a57a5.1427265330.git.chang-joon.lee@intel.com>
From: John Harrison <John.C.Harrison@Intel.com>
Date: Mon, 16 Mar 2015 14:26:43 +0000
Subject: [PATCH 18/22] SQUASHME! [VPG]: drm/i915: Fix for confused TDR when
 scheduler is in bypass mode

When the scheduler is in bypass mode, it does not keep track of any
state. Thus when the TDR code called in to ask if anything was
executing on a given ring, the answer returned was incorrect. This
patch adds bypass mode support to the query such that it now says that
the ring is (potentially) busy rather than definitely idle. This
allows the TDR code to keep going and properly investigate the ring.

The original commit that add the query function in question was
  commit 6a326e487ba93108c69dd501d3f85a62e2285ffc
  Author: John Harrison <John.C.Harrison@Intel.com>
  Date:   Wed Jan 14 15:40:37 2015 +0000

  FOR_UPSTREAM [VPG]: drm/i915: Add scheduler support functions for TDR

  Change-Id: I720463f01c4edd3579ce52e315a85e4d7874d7e5
  For: VIZ-1587
  For: VIZ-4741
  For: GMIN-3638
  Signed-off-by: John Harrison <John.C.Harrison@Intel.com>

This commit should be squashed in to that one.

Change-Id: I2565dbcdfbc6f53b9c5903f34a2f7a7055e9a858
For: VIZ-1587
For: VIZ-4741
Signed-off-by: John Harrison <John.C.Harrison@Intel.com>
---
 drivers/gpu/drm/i915/i915_scheduler.c |    5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_scheduler.c b/drivers/gpu/drm/i915/i915_scheduler.c
index 0a25b10..7bd7033 100644
--- a/drivers/gpu/drm/i915/i915_scheduler.c
+++ b/drivers/gpu/drm/i915/i915_scheduler.c
@@ -1675,6 +1675,11 @@ bool i915_scheduler_is_ring_flying(struct intel_engine_cs *ring)
 	unsigned long   flags;
 	bool            found = false;
 
+	/* With the scheduler in bypass mode, no information can be returned. */
+	if (i915.scheduler_override & i915_so_direct_submit) {
+		return true;
+	}
+
 	spin_lock_irqsave(&scheduler->lock, flags);
 
 	list_for_each_entry(node, &scheduler->node_queue[ring->id], link) {
-- 
1.7.9.5

